src/h3/mod.rs

```diff
@@ -26,4 +26,830 @@
+                .unwrap();
+
+            // TODO await ACK of stream open?
+            self.qpack_encoder_stream_open = true;
+        }
+
+        if !self.qpack_decoder_stream_open {
+            let mut d = [42; 128];
+            let mut dec_b = octets::Octets::with_slice(&mut d);
+            dec_b.put_u8(QPACK_DECODER_STREAM_TYPE_ID).unwrap();
+            let off = dec_b.off();
+            let stream_id = self.get_decoder_stream_id();
+            quic_conn
+                .stream_send(stream_id, &mut d[..off], false)
+                .unwrap();
+
+            // TODO await ACK of stream open?
+            self.qpack_decoder_stream_open = true;
+        }
+    }
+
+    pub fn create_placeholder_tree(&mut self, quic_conn: &mut super::Connection) {
+        if self.num_placeholders > 0 {
+            debug!("Going to prioritse {} placeholders", self.num_placeholders);
+            // TODO make sure slice is large enough to hold
+            // *all* PRIORITY frames. Worst case is ~7 bytes per frame.
+            let mut d = [42; 255];
+            let mut b = octets::Octets::with_slice(&mut d);
+
+            let mut weight = 0;
+            for i in 0..self.num_placeholders {
+                let frame = frame::H3Frame::Priority {
+                    priority_elem: frame::PrioritizedElemType::Placeholder,
+                    elem_dependency: frame::ElemDependencyType::RootOfTree,
+                    prioritized_element_id: Some(i),
+                    element_dependency_id: None,
+                    weight,
+                };
+
+                frame.to_bytes(&mut b).unwrap();
+
+                weight += 1;
+            }
+
+            let off = b.off();
+            debug!("Amount of priority bytes to send is {}", off);
+            let stream_id = self.get_control_stream_id();
+
+            quic_conn
+                .stream_send(stream_id, &mut d[..off], false)
+                .unwrap();
+        }
+    }
+
+    // Send SETTINGS frame based on H3 config
+    pub fn send_settings(&mut self, quic_conn: &mut super::Connection) {
+        self.open_control_stream(quic_conn);
+
+        let mut d = [42; 128];
+
+        let num_placeholders = if self.is_server { Some(16) } else { None };
+
+        let frame = frame::H3Frame::Settings {
+            num_placeholders,
+            max_header_list_size: Some(1024),
+            qpack_max_table_capacity: None,
+            qpack_blocked_streams: None,
+        };
+
+        let mut b = octets::Octets::with_slice(&mut d);
+
+        frame.to_bytes(&mut b).unwrap();
+        let off = b.off();
+        let stream_id = self.get_control_stream_id();
+
+        quic_conn
+            .stream_send(stream_id, &mut d[..off], false)
+            .unwrap();
+    }
+
+    // Send a request
+    pub fn send_request(
+        &mut self, request: http::Request<()>, quic_conn: &mut super::Connection,
+    ) {
+        let mut d = [42; 65535];
```

Comment:
We should find a better way to do this....

Reply:
Thinking about this some more, I think that using the `to_bytes()` method for serializing the frames isn't very good for this use-case.

So we should instead have a function that only write the frame header, call `stream_send()` with it, then call `stream_send()` again with encoded QPACK header block.

Reply:
I reworked these functions. They are 90% the same code anyway so that's been extracted out.